### PR TITLE
addressed PR feedbacks on object writer/reader

### PR DESCRIPTION
**Pull request by heejaechang**
also, some bug fixes on bit array I found whle using it in object writer/reader

### CODE REVIEW

**Comment by heejaechang** [comment]
@mattwar @tmat @dotnet/roslyn-compiler can you take a look?

**Comment by mattwar** [comment]
:+1:

File: src/Compilers/Core/CodeAnalysisTest/Collections/BitArrayTests.cs [hunk 1]
```diff
@@ -149,6 +150,65 @@ public void CheckTrueBits()
...
+            for (int capacity = 0; capacity < maxBits; capacity++)
+            {
+                BitVector b = BitVector.Create(capacity);
+                for (int i = 0; i < capacity; i++)
+                {
+                    b[i] = false;
+                }
+
+                var required = BitVector.WordsRequired(capacity);
+                var count = BitVector.AllSet(capacity).Words().Count();
```
Comment (cston) [inline src/Compilers/Core/CodeAnalysisTest/Collections/BitArrayTests.cs]: `Assert.Equal(count, required);`?

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs [hunk 2]
```diff
@@ -371,70 +371,93 @@ private Array ReadPrimitiveTypeArrayElements(Type type, DataKind kind, int lengt
...
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadUInt64);
+                case DataKind.Float4:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadSingle);
+                case DataKind.Float8:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadDouble);
+                case DataKind.Decimal:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadDecimal);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(kind);
+            }
```
Comment (cston) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: Consider moving the `switch` to a `GetPrimitiveElementRead` method and `return ReadPrimitiveTypeArrayElements(length, GetPrimitiveElementRead(_reader, kind));` here.

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs [hunk 3]
```diff
@@ -371,70 +371,93 @@ private Array ReadPrimitiveTypeArrayElements(Type type, DataKind kind, int lengt
...
+        }
+
+        private Array ReadBooleanArray(int length)
```
Comment (cston) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: `bool[]`?

Comment (cston) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: `private static bool[] ...`?

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs [hunk 4]
```diff
@@ -371,70 +371,93 @@ private Array ReadPrimitiveTypeArrayElements(Type type, DataKind kind, int lengt
...
+                    }
+
+                    array[count++] = BitVector.IsTrue(word, p);
                 }
             }
 
             return array;
         }
 
+        private T[] ReadPrimitiveTypeArrayElements<T>(int length, Func<T> read)
```
Comment (cston) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: `static`

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: `bool[]`

File: src/Compilers/Core/Portable/Serialization/ObjectWriter.cs [hunk 5]
```diff
@@ -514,55 +514,76 @@ private void WritePrimitiveTypeArrayElements(Type type, DataKind kind, Array ins
...
+            }
+
+            // send over bit array
+            foreach (var word in bits.Words())
+            {
+                _writer.Write(word);
+            }
+        }
+
+        private void WritePrimitiveTypeArrayElements<T>(T[] array, Action<T> write)
```
Comment (cston) [inline src/Compilers/Core/Portable/Serialization/ObjectWriter.cs]: `static`

**Pull request merged**